src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,212 @@
+ *  - Require PSM address in constructor
+ *  - The `push()` method swaps GEM to DAI using PSM
+ */
+contract RwaInputConduit3 {
+    /// @dev This is declared here so the storage layout lines up with RwaInputConduit.
+    DSTokenAbstract private __unused_gov;
+    /// @notice Dai token contract address
+    DSTokenAbstract public dai;
+    /// @notice RWA urn contract address
+    address public to;
+
+    /// @notice Addresses with admin access on this contract. `wards[usr]`
+    mapping(address => uint256) public wards;
+    /// @notice Addresses with push access on this contract. `may[usr]`
+    mapping(address => uint256) public may;
+
+    /// @notice PSM GEM token contract address
+    DSTokenAbstract immutable public gem;
+    /// @notice PSM contract address
+    PsmAbstract immutable public psm;
+    /// @notice Exit address
+    address public quitAddress;
+
+    /**
+     * @notice `usr` was granted admin access.
+     * @param usr The user address.
+     */
+    event Rely(address indexed usr);
+    /**
+     * @notice `usr` admin access was revoked.
+     * @param usr The user address.
+     */
+    event Deny(address indexed usr);
+    /**
+     * @notice `usr` was granted push access.
+     * @param usr The user address.
+     */
+    event Mate(address indexed usr);
+    /**
+     * @notice `usr` push access was revoked.
+     * @param usr The user address.
+     */
+    event Hate(address indexed usr);
+    /**
+     * @notice `wad` amount of Dai was pushed to `to`
+     * @param to The RwaUrn address
+     * @param wad The amount of Dai
+     */
+    event Push(address indexed to, uint256 wad);
+    /**
+     * @notice A contract parameter was updated.
+     * @param what The changed parameter name. Currently the supported values are: "quitAddress".
+     * @param data The new value of the parameter.
+     */
+    event File(bytes32 indexed what, address data);
+    /**
+     * @notice The conduit outstanding gem balance was flushed out to `exitAddress`.
+     * @param quitAddress The quiteAddress address.
+     * @param wad The amount flushed out.
+     */
+    event Quit(address indexed quitAddress, uint256 wad);
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "RwaInputConduit3/not-authorized");
+        _;
+    }
+
+    /**
+     * @notice Define addresses and gives `msg.sender` admin access.
+     * @param _psm PSM contract address.
+     * @param _to RwaUrn contract address.
+     */
+    constructor(address _psm, address _to, address _quitAddress) public {
+        DSTokenAbstract _gem = DSTokenAbstract(GemJoinAbstract(PsmAbstract(_psm).gemJoin()).gem());
+        psm = PsmAbstract(_psm);
+        dai = DSTokenAbstract(PsmAbstract(_psm).dai());
+        gem = _gem;
+        to = _to;
+        quitAddress = _quitAddress;
+
+        // Give unlimited approve to PSM gemjoin
+        _gem.approve(address(PsmAbstract(_psm).gemJoin()), 2**256 - 1);
```

Comment:
@Probably this way of writing it is more idiomatic and probably more gas efficient (2**256 does not fit into a single word, so unless the compiler is optimizing this away, it will take 2 words during execution).

```suggestion
        _gem.approve(address(PsmAbstract(_psm).gemJoin()), type(uint256).max);
```

Reply:
DOne